File: backend/app/schemas/notification.py
"""
通知管理Schema
"""
from typing import Optional, Dict, List, Any
from datetime import datetime
from pydantic import BaseModel, Field, validator
from enum import Enum

class NotificationType(str, Enum):
    """通知类型"""
    TRADE = "trade"
    RISK = "risk"
    SYSTEM = "system"
    MARKET = "market"
    ACCOUNT = "account"
    SECURITY = "security"

class NotificationChannel(str, Enum):
    """通知渠道"""
    IN_APP = "in_app"
    EMAIL = "email"
    SMS = "sms"
    PUSH = "push"
    WEBHOOK = "webhook"

class NotificationPriority(str, Enum):
    """通知优先级"""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    URGENT = "urgent"

class NotificationStatus(str, Enum):
    """通知状态"""
    PENDING = "pending"
    SENT = "sent"
    DELIVERED = "delivered"
    READ = "read"
    FAILED = "failed"
    CANCELLED = "cancelled"

# ==================== 通知模板 ====================

class NotificationTemplateBase(BaseModel):
    """通知模板基础模型"""
    name: str = Field(..., description="模板名称")
    code: str = Field(..., description="模板代码")
    type: NotificationType = Field(..., description="通知类型")
    title_template: str = Field(..., description="标题模板")
    content_template: str = Field(..., description="内容模板")
    channels: List[NotificationChannel] = Field(..., description="支持的通知渠道")
    variables: Optional[Dict[str, Any]] = Field(default_factory=dict, description="模板变量定义")
    default_priority: NotificationPriority = Field(default=NotificationPriority.NORMAL, description="默认优先级")
    is_active: bool = Field(default=True, description="是否激活")

class NotificationTemplateCreate(NotificationTemplateBase):
    """创建通知模板"""
    pass

class NotificationTemplateUpdate(BaseModel):
    """更新通知模板"""
    name: Optional[str] = None
    title_template: Optional[str] = None
    content_template: Optional[str] = None
    channels: Optional[List[NotificationChannel]] = None
    variables: Optional[Dict[str, Any]] = None
    default_priority: Optional[NotificationPriority] = None
    is_active: Optional[bool] = None

class NotificationTemplateResponse(NotificationTemplateBase):
    """通知模板响应"""
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

# ==================== 通知消息 ====================

class NotificationBase(BaseModel):
    """通知消息基础模型"""
    type: NotificationType = Field(..., description="通知类型")
    title: str = Field(..., description="通知标题")
    content: str = Field(..., description="通知内容")
    channel: NotificationChannel = Field(..., description="通知渠道")
    priority: NotificationPriority = Field(default=NotificationPriority.NORMAL, description="优先级")
    recipient: Optional[str] = Field(None, description="接收者")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="元数据")
    expires_at: Optional[datetime] = Field(None, description="过期时间")

class NotificationCreate(NotificationBase):
    """创建通知消息"""
    user_id: int = Field(..., description="用户ID")
    template_id: Optional[int] = Field(None, description="模板ID")
    variables: Optional[Dict[str, Any]] = Field(default_factory=dict, description="模板变量值")

class NotificationUpdate(BaseModel):
    """更新通知消息"""
    status: Optional[NotificationStatus] = None
    read_at: Optional[datetime] = None

class NotificationResponse(NotificationBase):
    """通知消息响应"""
    id: int
    user_id: int
    template_id: Optional[int]
    status: NotificationStatus
    sent_at: Optional[datetime]
    delivered_at: Optional[datetime]
    read_at: Optional[datetime]
    error_message: Optional[str]
    retry_count: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

class NotificationBatchCreate(BaseModel):
    """批量创建通知"""
    user_ids: List[int] = Field(..., description="用户ID列表")
    template_code: str = Field(..., description="模板代码")
    variables: Optional[Dict[str, Any]] = Field(default_factory=dict, description="模板变量")
    channels: List[NotificationChannel] = Field(..., description="通知渠道")
    priority: NotificationPriority = Field(default=NotificationPriority.NORMAL, description="优先级")
    scheduled_at: Optional[datetime] = Field(None, description="计划发送时间")

class NotificationBatchResponse(BaseModel):
    """批量通知响应"""
    total_count: int = Field(..., description="总数量")
    success_count: int = Field(..., description="成功数量")
    failed_count: int = Field(..., description="失败数量")
    notification_ids: List[int] = Field(..., description="通知ID列表")

# ==================== 通知偏好 ====================

class NotificationPreferenceBase(BaseModel):
    """通知偏好基础模型"""
    enabled: bool = Field(default=True, description="是否启用通知")
    quiet_hours_enabled: bool = Field(default=False, description="是否启用免打扰时间")
    quiet_hours_start: str = Field(default="22:00", description="免打扰开始时间")
    quiet_hours_end: str = Field(default="08:00", description="免打扰结束时间")
    email_enabled: bool = Field(default=True, description="邮件通知启用")
    sms_enabled: bool = Field(default=False, description="短信通知启用")
    push_enabled: bool = Field(default=True, description="推送通知启用")
    in_app_enabled: bool = Field(default=True, description="站内信启用")
    trade_notifications: Optional[Dict[str, bool]] = Field(default_factory=dict, description="交易通知偏好")
    risk_notifications: Optional[Dict[str, bool]] = Field(default_factory=dict, description="风险通知偏好")
    system_notifications: Optional[Dict[str, bool]] = Field(default_factory=dict, description="系统通知偏好")
    market_notifications: Optional[Dict[str, bool]] = Field(default_factory=dict, description="市场通知偏好")
    account_notifications: Optional[Dict[str, bool]] = Field(default_factory=dict, description="账户通知偏好")
    security_notifications: Optional[Dict[str, bool]] = Field(default_factory=dict, description="安全通知偏好")
    max_notifications_per_hour: int = Field(default=10, description="每小时最大通知数")
    digest_enabled: bool = Field(default=False, description="是否启用摘要通知")
    digest_frequency: str = Field(default="daily", description="摘要频率")
    digest_time: str = Field(default="09:00", description="摘要发送时间")

    @validator('quiet_hours_start', 'quiet_hours_end', 'digest_time')
    def validate_time_format(cls, v):
        """验证时间格式"""
        if v and not v.match(r'^([01]?[0-9]|2[0-3]):[0-5][0-9]$'):
            raise ValueError('时间格式必须为 HH:MM')
        return v

class NotificationPreferenceUpdate(NotificationPreferenceBase):
    """更新通知偏好"""
    pass

class NotificationPreferenceResponse(NotificationPreferenceBase):
    """通知偏好响应"""
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

# ==================== 通知规则 ====================

class NotificationRuleBase(BaseModel):
    """通知规则基础模型"""
    name: str = Field(..., description="规则名称")
    description: Optional[str] = Field(None, description="规则描述")
    event_type: str = Field(..., description="事件类型")
    conditions: Dict[str, Any] = Field(..., description="触发条件")
    template_code: Optional[str] = Field(None, description="模板代码")
    channels: List[NotificationChannel] = Field(..., description="通知渠道")
    priority: NotificationPriority = Field(default=NotificationPriority.NORMAL, description="优先级")
    rate_limit: Optional[int] = Field(None, description="频率限制（分钟）")
    max_per_day: Optional[int] = Field(None, description="每日最大次数")
    is_active: bool = Field(default=True, description="是否激活")

class NotificationRuleCreate(NotificationRuleBase):
    """创建通知规则"""
    pass

class NotificationRuleUpdate(BaseModel):
    """更新通知规则"""
    name: Optional[str] = None
    description: Optional[str] = None
    conditions: Optional[Dict[str, Any]] = None
    template_code: Optional[str] = None
    channels: Optional[List[NotificationChannel]] = None
    priority: Optional[NotificationPriority] = None
    rate_limit: Optional[int] = None
    max_per_day: Optional[int] = None
    is_active: Optional[bool] = None

class NotificationRuleResponse(NotificationRuleBase):
    """通知规则响应"""
    id: int
    user_id: int
    trigger_count: int
    last_triggered_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

# ==================== 通知统计 ====================

class NotificationStats(BaseModel):
    """通知统计"""
    total_notifications: int = Field(..., description="总通知数")
    unread_notifications: int = Field(..., description="未读通知数")
    notifications_by_type: Dict[str, int] = Field(..., description="按类型统计")
    notifications_by_channel: Dict[str, int] = Field(..., description="按渠道统计")
    notifications_by_status: Dict[str, int] = Field(..., description="按状态统计")
    recent_notifications: List[NotificationResponse] = Field(..., description="最近通知")

class NotificationDigestResponse(BaseModel):
    """通知摘要响应"""
    id: int
    user_id: int
    digest_type: str
    period_start: datetime
    period_end: datetime
    total_notifications: int
    unread_notifications: int
    notification_summary: Dict[str, Any]
    is_sent: bool
    sent_at: Optional[datetime]
    created_at: datetime
    
    class Config:
        orm_mode = True

# ==================== 请求模型 ====================

class NotificationMarkReadRequest(BaseModel):
    """标记已读请求"""
    notification_ids: List[int] = Field(..., description="通知ID列表")

class NotificationBatchDeleteRequest(BaseModel):
    """批量删除请求"""
    notification_ids: List[int] = Field(..., description="通知ID列表")
    delete_type: str = Field(default="soft", description="删除类型: soft/hard")

class NotificationTestRequest(BaseModel):
    """测试通知请求"""
    template_code: str = Field(..., description="模板代码")
    channel: NotificationChannel = Field(..., description="通知渠道")
    variables: Optional[Dict[str, Any]] = Field(default_factory=dict, description="模板变量")
    recipient: Optional[str] = Field(None, description="接收者")

class NotificationSearchRequest(BaseModel):
    """通知搜索请求"""
    keyword: Optional[str] = Field(None, description="关键词")
    type: Optional[NotificationType] = Field(None, description="通知类型")
    channel: Optional[NotificationChannel] = Field(None, description="通知渠道")
    status: Optional[NotificationStatus] = Field(None, description="通知状态")
    priority: Optional[NotificationPriority] = Field(None, description="优先级")
    date_from: Optional[datetime] = Field(None, description="开始日期")
    date_to: Optional[datetime] = Field(None, description="结束日期")
    page: int = Field(default=1, description="页码")
    page_size: int = Field(default=20, description="每页数量")

class NotificationSearchResponse(BaseModel):
    """通知搜索响应"""
    notifications: List[NotificationResponse] = Field(..., description="通知列表")
    total_count: int = Field(..., description="总数量")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")
    total_pages: int = Field(..., description="总页数")

# ==================== 通知设置 ====================

class NotificationChannelConfig(BaseModel):
    """通知渠道配置"""
    channel: NotificationChannel = Field(..., description="通知渠道")
    enabled: bool = Field(default=True, description="是否启用")
    config: Dict[str, Any] = Field(default_factory=dict, description="渠道配置")

class NotificationSystemConfig(BaseModel):
    """通知系统配置"""
    channels: List[NotificationChannelConfig] = Field(..., description="渠道配置")
    default_retry_count: int = Field(default=3, description="默认重试次数")
    retry_intervals: List[int] = Field(default=[60, 300, 900], description="重试间隔（秒）")
    cleanup_days: int = Field(default=30, description="清理天数")
    rate_limit_enabled: bool = Field(default=True, description="是否启用频率限制")
    batch_size: int = Field(default=100, description="批处理大小")

# ==================== 导出模型 ====================

class NotificationExportRequest(BaseModel):
    """通知导出请求"""
    export_type: str = Field(default="notifications", description="导出类型")
    filters: Optional[NotificationSearchRequest] = Field(None, description="过滤条件")
    format: str = Field(default="csv", description="导出格式")
    include_content: bool = Field(default=True, description="是否包含内容")

class NotificationExportResponse(BaseModel):
    """通知导出响应"""
    export_id: str = Field(..., description="导出ID")
    status: str = Field(..., description="导出状态")
    download_url: Optional[str] = Field(None, description="下载链接")
    file_size: Optional[int] = Field(None, description="文件大小")
    expires_at: Optional[datetime] = Field(None, description="过期时间")
    created_at: datetime = Field(..., description="创建时间")

File: frontend/src/api/notification.ts
/**
 * 通知管理API接口
 */
import { request } from '@/utils/request'

export enum NotificationType {
  TRADE = 'trade',
  RISK = 'risk',
  SYSTEM = 'system',
  MARKET = 'market',
  ACCOUNT = 'account',
  SECURITY = 'security'
}

export enum NotificationChannel {
  IN_APP = 'in_app',
  EMAIL = 'email',
  SMS = 'sms',
  PUSH = 'push',
  WEBHOOK = 'webhook'
}

export enum NotificationPriority {
  LOW = 'low',
  NORMAL = 'normal',
  HIGH = 'high',
  URGENT = 'urgent'
}

export enum NotificationStatus {
  PENDING = 'pending',
  SENT = 'sent',
  DELIVERED = 'delivered',
  READ = 'read',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

export interface Notification {
  id: number
  user_id: number
  template_id?: number
  type: NotificationType
  title: string
  content: string
  channel: NotificationChannel
  priority: NotificationPriority
  status: NotificationStatus
  recipient?: string
  sent_at?: string
  delivered_at?: string
  read_at?: string
  error_message?: string
  retry_count: number
  metadata?: Record<string, any>
  expires_at?: string
  created_at: string
  updated_at: string
}

export interface NotificationTemplate {
  id: number
  name: string
  code: string
  type: NotificationType
  title_template: string
  content_template: string
  channels: NotificationChannel[]
  variables: Record<string, any>
  default_priority: NotificationPriority
  created_at: string
  updated_at: string
}

export interface NotificationPreference {
  id: number
  user_id: number
  enabled: boolean
  quiet_hours_enabled: boolean
  quiet_hours_start: string
  quiet_hours_end: string
  email_enabled: boolean
  sms_enabled: boolean
  push_enabled: boolean
  in_app_enabled: boolean
  trade_notifications: Record<string, boolean>
  risk_notifications: Record<string, boolean>
  system_notifications: Record<string, boolean>
  market_notifications: Record<string, boolean>
  account_notifications: Record<string, boolean>
  security_notifications: Record<string, boolean>
  max_notifications_per_hour: number
  digest_enabled: boolean
  digest_frequency: string
  digest_time: string
  created_at: string
  updated_at: string
}

export interface NotificationRule {
  id: number
  user_id: number
  name: string
  description?: string
  event_type: string
  conditions: Record<string, any>
  template_code?: string
  channels: NotificationChannel[]
  priority: NotificationPriority
  rate_limit?: number
  max_per_day?: number
  is_active: boolean
  trigger_count: number
  last_triggered_at?: string
  created_at: string
  updated_at: string
}

export interface NotificationStats {
  total_notifications: number
  unread_notifications: number
  notifications_by_type: Record<string, number>
  notifications_by_channel: Record<string, number>
  notifications_by_status: Record<string, number>
  recent_notifications: Notification[]
}

export interface CreateNotificationRequest {
  user_id: number
  template_id?: number
  type: NotificationType
  title: string
  content: string
  channel: NotificationChannel
  priority?: NotificationPriority
  recipient?: string
  metadata?: Record<string, any>
  variables?: Record<string, any>
  expires_at?: string
}

export interface BatchCreateNotificationRequest {
  user_ids: number[]
  template_code: string
  variables?: Record<string, any>
  channels: NotificationChannel[]
  priority?: NotificationPriority
  scheduled_at?: string
}

export interface NotificationSearchRequest {
  keyword?: string
  type?: NotificationType
  channel?: NotificationChannel
  status?: NotificationStatus
  priority?: NotificationPriority
  date_from?: string
  date_to?: string
  page?: number
  page_size?: number
}

export interface NotificationSearchResponse {
  notifications: Notification[]
  total_count: number
  page: number
  page_size: number
  total_pages: number
}

// ==================== 通知消息管理 ====================

// 获取用户通知列表
export const getUserNotifications = (params: {
  limit?: number
  skip?: number
  unread_only?: boolean
}): Promise<{
  notifications: Notification[]
  total_count: number
  unread_count: number
  page_info: {
    skip: number
    limit: number
    has_more: boolean
  }
}> => {
  return request.get('/v1/notifications/', { params })
}

// 创建通知
export const createNotification = (data: CreateNotificationRequest): Promise<Notification> => {
  return request.post('/v1/notifications/', data)
}

// 批量创建通知
export const batchCreateNotifications = (data: BatchCreateNotificationRequest): Promise<{
  total_count: number
  success_count: number
  failed_count: number
  notification_ids: number[]
}> => {
  return request.post('/v1/notifications/batch', data)
}

// 搜索通知
export const searchNotifications = (data: NotificationSearchRequest): Promise<NotificationSearchResponse> => {
  return request.post('/v1/notifications/search', data)
}

// 标记通知为已读
export const markNotificationsRead = (notification_ids: number[]): Promise<{ message: string }> => {
  return request.post('/v1/notifications/mark-read', { notification_ids })
}

// 标记所有通知为已读
export const markAllNotificationsRead = (): Promise<{ message: string }> => {
  return request.post('/v1/notifications/mark-all-read')
}

// 批量删除通知
export const deleteNotifications = (notification_ids: number[], delete_type: 'soft' | 'hard' = 'soft'): Promise<{ message: string }> => {
  return request.delete('/v1/notifications/', {
    data: { notification_ids, delete_type }
  })
}

// 删除所有通知
export const deleteAllNotifications = (delete_type: 'soft' | 'hard' = 'soft'): Promise<{ message: string }> => {
  return request.delete('/v1/notifications/all', {
    params: { delete_type }
  })
}

// ==================== 通知偏好管理 ====================

// 获取通知偏好设置
export const getNotificationPreferences = (): Promise<NotificationPreference> => {
  return request.get('/v1/notifications/preferences')
}

// 更新通知偏好设置
export const updateNotificationPreferences = (data: Partial<NotificationPreference>): Promise<NotificationPreference> => {
  return request.put('/v1/notifications/preferences', data)
}

// ==================== 通知规则管理 ====================

// 创建通知规则
export const createNotificationRule = (data: Omit<NotificationRule, 'id' | 'user_id' | 'trigger_count' | 'last_triggered_at' | 'created_at' | 'updated_at'>): Promise<NotificationRule> => {
  return request.post('/v1/notifications/rules', data)
}

// 获取通知规则列表
export const getNotificationRules = (): Promise<NotificationRule[]> => {
  return request.get('/v1/notifications/rules')
}

// 更新通知规则
export const updateNotificationRule = (rule_id: number, data: Partial<NotificationRule>): Promise<NotificationRule> => {
  return request.put(`/v1/notifications/rules/${rule_id}`, data)
}

// 删除通知规则
export const deleteNotificationRule = (rule_id: number): Promise<{ message: string }> => {
  return request.delete(`/v1/notifications/rules/${rule_id}`)
}

// ==================== 通知模板管理 ====================

// 获取通知模板列表
export const getNotificationTemplates = (type_filter?: NotificationType): Promise<NotificationTemplate[]> => {
  return request.get('/v1/notifications/templates', {
    params: type_filter ? { type_filter } : {}
  })
}

// 获取通知模板详情
export const getNotificationTemplate = (template_code: string): Promise<NotificationTemplate> => {
  return request.get(`/v1/notifications/templates/${template_code}`)
}

// 测试通知模板
export const testNotificationTemplate = (data: {
  template_code: string
  channel: NotificationChannel
  variables?: Record<string, any>
  recipient?: string
}): Promise<any> => {
  return request.post('/v1/notifications/templates/test', data)
}

// ==================== 通知统计 ====================

// 获取通知统计
export const getNotificationStats = (days: number = 30): Promise<NotificationStats> => {
  return request.get('/v1/notifications/stats', {
    params: { days }
  })
}

// 获取未读通知数量
export const getUnreadCount = (): Promise<{ unread_count: number }> => {
  return request.get('/v1/notifications/unread-count')
}

// ==================== 通知导出 ====================

// 导出通知数据
export const exportNotifications = (data: {
  export_type?: string
  filters?: NotificationSearchRequest
  format?: string
  include_content?: boolean
}): Promise<{
  export_id: string
  status: string
  download_url?: string
  file_size?: number
  expires_at?: string
  created_at: string
}> => {
  return request.post('/v1/notifications/export', data)
}

// ==================== 工具函数 ====================

// 通知类型显示名称
export const getNotificationTypeLabel = (type: NotificationType): string => {
  const labels = {
    [NotificationType.TRADE]: '交易通知',
    [NotificationType.RISK]: '风险通知',
    [NotificationType.SYSTEM]: '系统通知',
    [NotificationType.MARKET]: '市场通知',
    [NotificationType.ACCOUNT]: '账户通知',
    [NotificationType.SECURITY]: '安全通知'
  }
  return labels[type] || type
}

// 通知渠道显示名称
export const getNotificationChannelLabel = (channel: NotificationChannel): string => {
  const labels = {
    [NotificationChannel.IN_APP]: '站内信',
    [NotificationChannel.EMAIL]: '邮件',
    [NotificationChannel.SMS]: '短信',
    [NotificationChannel.PUSH]: '推送',
    [NotificationChannel.WEBHOOK]: 'Webhook'
  }
  return labels[channel] || channel
}

// 通知优先级显示名称
export const getNotificationPriorityLabel = (priority: NotificationPriority): string => {
  const labels = {
    [NotificationPriority.LOW]: '低',
    [NotificationPriority.NORMAL]: '普通',
    [NotificationPriority.HIGH]: '高',
    [NotificationPriority.URGENT]: '紧急'
  }
  return labels[priority] || priority
}

// 通知状态显示名称
export const getNotificationStatusLabel = (status: NotificationStatus): string => {
  const labels = {
    [NotificationStatus.PENDING]: '待发送',
    [NotificationStatus.SENT]: '已发送',
    [NotificationStatus.DELIVERED]: '已送达',
    [NotificationStatus.READ]: '已读',
    [NotificationStatus.FAILED]: '发送失败',
    [NotificationStatus.CANCELLED]: '已取消'
  }
  return labels[status] || status
}

// 获取通知优先级颜色
export const getNotificationPriorityColor = (priority: NotificationPriority): string => {
  const colors = {
    [NotificationPriority.LOW]: '#909399',
    [NotificationPriority.NORMAL]: '#409eff',
    [NotificationPriority.HIGH]: '#e6a23c',
    [NotificationPriority.URGENT]: '#f56c6c'
  }
  return colors[priority] || colors[NotificationPriority.NORMAL]
}

// 获取通知状态颜色
export const getNotificationStatusColor = (status: NotificationStatus): string => {
  const colors = {
    [NotificationStatus.PENDING]: '#e6a23c',
    [NotificationStatus.SENT]: '#409eff',
    [NotificationStatus.DELIVERED]: '#67c23a',
    [NotificationStatus.READ]: '#909399',
    [NotificationStatus.FAILED]: '#f56c6c',
    [NotificationStatus.CANCELLED]: '#c0c4cc'
  }
  return colors[status] || colors[NotificationStatus.PENDING]
}

// 格式化通知时间
export const formatNotificationTime = (time: string): string => {
  const date = new Date(time)
  const now = new Date()
  const diffMs = now.getTime() - date.getTime()
  const diffMins = Math.floor(diffMs / (1000 * 60))
  const diffHours = Math.floor(diffMs / (1000 * 60 * 60))
  const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24))

  if (diffMins < 1) return '刚刚'
  if (diffMins < 60) return `${diffMins}分钟前`
  if (diffHours < 24) return `${diffHours}小时前`
  if (diffDays < 7) return `${diffDays}天前`
  
  return date.toLocaleDateString('zh-CN', {
    year: 'numeric',
    month: 'short',
    day: 'numeric'
  })
}

// 检查通知是否已过期
export const isNotificationExpired = (notification: Notification): boolean => {
  if (!notification.expires_at) return false
  return new Date(notification.expires_at) < new Date()
}

// 检查通知是否未读
export const isNotificationUnread = (notification: Notification): boolean => {
  return !notification.read_at
}